cucumber-expressions/java/src/main/java/io/cucumber/cucumberexpressions/ParameterTypeMatcher.java

```diff
@@ -1,30 +1,46 @@
 package io.cucumber.cucumberexpressions;
 
 import java.util.regex.Matcher;
+import java.util.regex.Pattern;
 
 class ParameterTypeMatcher implements Comparable<ParameterTypeMatcher> {
     private final ParameterType<?> parameterType;
     private final Matcher matcher;
-    private final int textLength;
+    private final String text;
 
-    public ParameterTypeMatcher(ParameterType<?> parameterType, Matcher matcher, int textLength) {
+    public ParameterTypeMatcher(ParameterType<?> parameterType, Matcher matcher, String text) {
         this.parameterType = parameterType;
         this.matcher = matcher;
-        this.textLength = textLength;
+        this.text = text;
     }
 
     public boolean advanceToAndFind(int newMatchPos) {
         // Unlike js, ruby and go, the matcher is stateful
         // so we can't use the immutable semantics.
-        matcher.region(newMatchPos, textLength);
+        matcher.region(newMatchPos, text.length());
         while (matcher.find()) {
             if (!group().isEmpty()) {
+                if (matcher.start() > 0) {
+                    char before = text.charAt(matcher.start() - 1);
+                    if (!isWhitespaceOrPunctuation(before)) {
+                        return false;
+                    }
+                }
+
+                if (matcher.end() < text.length()) {
+                    char after = text.charAt(matcher.end());
+                    return isWhitespaceOrPunctuation(after);
+                }
```

Comment:
I find this section really hard to follow. Perhaps we extract a method here that makes it clear we should have a match and that the match should be a full word.

Reply:
So something like 
```java
             if (!group().isEmpty() && matchIsFullWord()) { 
```